krun/krun.c

```diff
@@ -0,0 +1,238 @@
+ *
+ * Copyright (C) 2017, King's College London.
+ * Authored by Edd Barrett <vext01@gmail.com>.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#include <linux/kernel.h>
+#include <linux/types.h>
+#include <linux/errno.h>
+#include <linux/smp.h>
+#include <asm/msr.h>
+#include <linux/uaccess.h>
+
+#include "krun_reg.h"
+
+/* protos */
+void krun_read_msrs(void *data);
+void krun_reset_msrs(void *ununsed);
+asmlinkage int sys_krun_read_msrs(int n_cores, bool ctr1_first, u64 *aperfs,
+    u64 *mperfs, u64 *ctr1s);
+asmlinkage void sys_krun_reset_msrs(int n_cores);
+asmlinkage int sys_krun_configure(int n_cores);
+u8 krun_get_arch_perf_ctr_version(void);
+
+/* For passing args to krun_read_msrs() */
+struct krun_msr_args {
+	u64 aperf;
+	u64 mperf;
+	u64 ctr1;
+	bool ctr1_first;
+};
+
+/*
+ * Take low-latency MSR measurements for the current CPU core.
+ */
+void krun_read_msrs(void *data)
+{
+	u32 hi, lo;
+	struct krun_msr_args *args = (struct krun_msr_args *) data;
+
+	if (args->ctr1_first) {
+		asm volatile(
+		    "rdmsr\n\t"
+		    : "=d" (hi), "=a" (lo)	// out
+		    : "c"(IA32_PERF_FIXED_CTR1) // in
+		    :);				// clobber
+		args->ctr1 = ((u64) hi << 32) | lo;
+	}
+
+	asm volatile(
+	    "rdmsr\n\t"
+	    : "=d" (hi), "=a" (lo)	// out
+	    : "c"(IA32_APERF)		// in
+	    :);				// clobber
+	args->aperf = ((u64) hi << 32) | lo;
+
+	asm volatile(
+	    "rdmsr\n\t"
+	    : "=d" (hi), "=a" (lo)	// out
+	    : "c"(IA32_MPERF)		// in
+	    : );			// clobber
+	args->mperf = ((u64) hi << 32) | lo;
+
+	if (!args->ctr1_first) {
+		asm volatile(
+		    "rdmsr\n\t"
+		    : "=d" (hi), "=a" (lo)	// out
+		    : "c"(IA32_PERF_FIXED_CTR1) // in
+		    :);				// clobber
+		args->ctr1 = ((u64) hi << 32) | lo;
+	}
+}
+
+/* Reset the MSRs of interest on the current CPU core */
+void krun_reset_msrs(void *ununsed)
+{
+	/*
+	 * Note we reset MPERF before APERF, so that APERF can never be greater
+	 * than MPERF.
+	 */
+	asm volatile(
+	    "xor %%eax, %%eax\n\t"
+	    "xor %%edx, %%edx\n\t"
+	    "mov %0, %%ecx\n\t"
+	    "wrmsr\n\t"
+	    "mov %1, %%ecx\n\t"
+	    "wrmsr\n\t"
+	    "mov %2, %%ecx\n\t"
+	    "wrmsr\n\t"
+	    :				// out
+	    : "i"(IA32_MPERF), "i"(IA32_APERF), "i"(IA32_PERF_FIXED_CTR1) // in
+	    :"eax", "ecx", "edx");	// clobber
+}
+
+u8
```

Comment:
I think this is the only function where the return type is on the line above the function name?

Reply:
oops. I was in BSD mode then apparently.